man/ola_rdm_discover.1
.TH ola_rdm_discover 1 "October 2015"
.SH NAME
ola_rdm_discover \- Fetch the UID list for a universe.
.SH SYNOPSIS
.B ola_rdm_discover
--universe <universe> [--full|--incremental]
.SH DESCRIPTION
.B ola_rdm_discover
Fetch the UID list for a universe.
.SH OPTIONS
.IP "-f, --full"
Force full RDM Discovery for this universe
.IP "-h, --help"
Display the help message
.IP "-i, --incremental"
Force incremental RDM Discovery for this universe
.IP "-l, --log-level <int8_t>"
Set the logging level 0 .. 4.
.IP "-u, --universe <uint32_t>"
The universe to do RDM discovery on
.IP "-v, --version"
Display version information
.IP "--include-broadcast"
Include broadcast UID for this universe
.IP "--include-vendorcast"
Include vendorcast UID for this universe
.IP "--no-use-epoll"
Disable the use of epoll(), revert to select()
.IP "--scheduler-policy <string>"
The thread scheduling policy, one of {fifo, rr}.
.IP "--scheduler-priority <uint16_t>"
The thread priority, only used if --scheduler-policy is set.
.IP "--syslog"
Send to syslog rather than stderr.


NEWS
x/y/2015 ola-0.9.9
 Features:
 * Add better logging for open()
 * Added more man pages.

 API:
 * 

 RDM Tests:
 * 

 Bugs:
 * 

 Internal:
 *

30/9/2015 ola-0.9.8
 Features:
 * Add support for the DMXter4A
 * Tidy the output of the Python ola_rdm_get for supported_parameters

 RDM Tests:
 * Fix a bug related to split levels not being supported in the dimmer PIDs
 * Allow NR_UNSUPPORTED_COMMAND_CLASS for SET DMX_ADDRESS when the address is
   not required
 * Switch from a warning to an advisory if it's a manufacturer specific use in
   GetSensorDefinition
 * Fix some other minor RDM test bugs

 Bugs:
 * Fix some Xcode build errors
 * Fix the build on Windows
 * Fix a bug in Artnet universe mapping #856
 * NACK requests for IPv4 default route PID with data
 * NACK with format error requests for DNS PIDs with data
 * Fix a bug regarding stack traces error messages with Python RDM gets #884
 * Support 0.9.43 of microhttpd

13/7/2015 ola-0.9.7
 Bugs:
 * RDM NACK codes displayed with incorrect byte ordering in the C++ client #812
 * Fix endian issues on s390x and PPC architectures.
 * Fix ShowNet endian issues.
 * Add some missing files to the tarball.
 * Restore binary compatability with 0.9.5 #826

30/6/2015 ola-0.9.6
 Features:
 * Default to using epoll() rather than select() if it exists,
 * Default to the async libusb implementation in the usbdmx plugin.
 * Added support for the APA102 family of SPI pixels.
 * Allow plugins to be enabled/disabled from the new web UI & CLI tools #682.
 * Add an option to display the frame and timing data in the RDM CLI tools
   where available.
 * Add additional C++ client documentation and examples.
 * Support an overrides.proto file to load system specific PID data #732.
 * Respect UUCP lock files before opening devices #716.
 * Initial partial E1.37-2 support in the web UI.

 API:
 * Python: Fix a bug where callbacks weren't run if the operation failed.
 * Python: improve Python 3 compatability.
 * Add patched ports in Universe protobuf message.
 * Python: Add a patch/unpatch example.
 * Python Add GetCandidatePorts and an example.

 RDM Tests:
 * Initial E1.37-2 tests.
 * Fix a bug in the SetDeviceHours test for devices that NAck set DEVICE_HOURS
   with a NR_WRITE_PROTECT.

 Bugs:
 * Fix port with unset universe appearing as bound to universe 0 in the Python
   client.
 * Ensure universe names and LTP priorities are always written to the config
   file #661.
 * Stop tests from failing when no non-loopback interfaces are present #763
   Debian #769670.
 * Fix various configure script behaviour issues #773.
 * Fix build issues with GCC-5 #765 Debian #778038.
 * Restore LIBS when we can't find the Saleae library during configure.
 * Fix the async libusb Eurolite Pro implementation.

 Internal:
 * Rename the method Source to ConfigLocation for Preference classes and
   associated changes.
 * Major cleanup of RDM code.
 * Switch to using pthread_equal rather than operator==.
 * Switch to using Dot for Doxygen diagrams where present.

1/3/2015 ola-0.9.5
 Features:
 * Work on a new Web-UI which is built with AngularJS
 * Added E1.37-2 support to the SPI plugin RDM responder
 * USB DMX plug now works on Windows (Scanlime Fadecandy tested working)
 * Update the deb package's udev rules

 Bugs:
 * Fix finding ola_protoc_plugin when cross-compiling.
 * Fix ncurses flags issues between configure and the Makefile.
 * Fix some more build issues on Windows XP
 * Fix our E1.37-2 implementation to match the E1.37-2 standard.
 * Fix the old website's RDM patcher identify function

29/1/2015 ola-0.9.4
 Features:
 * Added support for hotplugging USB devices using the usbdmx plugin, #374.
 * Added support for setting the thread scheduling options, #399.
 * Added support for the Open Pixel Control protocol, #502.
 * Added support for the Scanlime Fadecandy device, #503.
 * Added a GPIO Plugin, closes #299.
 * Added support for the Open DMX clone from KWMATIK.

 API:
 * Plugin libraries that were previously installed under $PREFIX/lib/olad/ are
   now in $PREFIX/lib/

 RDM Tests:
 * Fix a bug with the GetPresetInfo test around display multipliers.

 Bugs:
 * Add a workaround to make the Windows HTTP server more responsive.
 * Fixed defects detected by Coverity Scan.

 Internal:
 * Remove the SLP code since it's no longer required for E1.33.

16/11/2014 ola-0.9.3
 Features:
 * Send the E1.31 Stream Terminated message on shutdown and/or port unpatching.
 * Add support for the draft E1.31 discovery protocol.
 * Switch the dummy responder which implements the E1.37-2 PIDs to their final
   values.
 * Add the ability to whitelist rather than blacklist plugins during configure
   (via --disable-all-plugins) and switch the Windows build to this method.
 * Convert ola_trigger to our flags module; change it's default universe value
   to be 0 to align with other programs.
 * Make the ola_trigger config syntax more flexible, including dots in program
   names and quoted program names #507
 * Add an instance name to OLA server and use it within E1.31 and the DNS-SD
   code to make multiple instances of OLA easier to track. #473

 Bugs:
 * Fix some build issues on OS X 10.5.8 #493
 * Fix some build issues on Windows XP
 * Use the correct ifdefs within plugin loading so we only load stuff if we
   asked for it
 * Fix the inability to Ctrl+C in ola_recorder

 API:
 * Update ola::network::IPV4Address and ola::network::IPV4SocketAddress so they
   sort the same regardless of endianness and in a logical way for humans.

 Internal:
 * Update the configure version method to ensure the version files are updated
   automatically when it changes.
 * Add a generic RpcServer class and switch OlaServer to use it.
 * SelectServer now runs any remaining pending callbacks during shutdown.

20/8/2014 ola-0.9.2
 Features:
 * Added --use-epoll which uses epoll() rather than select().
 * Added --use-kqueue which uses kqueue() rather than select().
 * Add some more additional slot definitions.

python/examples/ola_rdm_discover.py
#!/usr/bin/env python
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#
# ola_rdm_discover.py
# Copyright (C) 2010 Simon Newton

'''Show the UIDs for a universe.'''

__author__ = 'nomis52@gmail.com (Simon Newton)'

import getopt
import textwrap
import sys
from ola.ClientWrapper import ClientWrapper


def Usage():
  print(textwrap.dedent("""\
  Usage: ola_rdm_discover.py --universe <universe> [--force_discovery]

  Fetch the UID list for a universe.

    -h, --help                Display this help message and exit.
    -f, --full                Full RDM Discovery for this universe.
    -i, --incremental         Incremental RDM Discovery for this universe.
    -u, --universe <universe> Universe number."""))


def main():
  try:
      opts, args = getopt.getopt(sys.argv[1:], 'fjiu:',
                                 ['help', 'full', 'incremental', 'universe='])
  except getopt.GetoptError as err:
    print(str(err))
    Usage()
    sys.exit(2)

  universe = None
  full_discovery = False
  incremental_discovery = False

  for o, a in opts:
    if o in ('-h', '--help'):
      Usage()
      sys.exit()
    elif o in ('-f', '--full'):
      full_discovery = True
    elif o in ('-i', '--incremental'):
      incremental_discovery = True
    elif o in ('-u', '--universe'):
      universe = int(a)

  if not universe:
    Usage()
    sys.exit()

  if incremental_discovery and full_discovery:
    print('Only one of --incremental or --full can be specified')
    sys.exit()

  wrapper = ClientWrapper()
  client = wrapper.Client()

  def show_uids(state, uids):
    if state.Succeeded():
      for uid in uids:
        print(str(uid))
    wrapper.Stop()

  if full_discovery:
    client.RunRDMDiscovery(universe, True, show_uids)
  elif incremental_discovery:
    client.RunRDMDiscovery(universe, False, show_uids)
  else:
    client.FetchUIDList(universe, show_uids)
  wrapper.Run()

if __name__ == '__main__':
  main()
